### PR TITLE
Cacheing

### PR DESCRIPTION
**Pull request by martindurant**

### CODE REVIEW

**Comment by martindurant** [comment]
Not sure how to best test this...

**Comment by martindurant** [comment]
@rabernat , you should find this interesting. Not entirely sure of the right usage pattern for something like this.

**Comment by martindurant** [comment]
(or perhaps you should wait until I add docstrings)

**Comment by martindurant** [comment]
Fixes #48

**Comment by rabernat** [comment]
I am very interested in this.

Does it only cache full files, or can it cache byte ranges?

**Comment by martindurant** [comment]
It caches *blocks* in sparse files. How much space that needs remains to be seen.

**Comment by rabernat** [comment]
What happens to the cache if you pass the filesystem through a distributed dask cluster. Is the whole thing pickled up? Or is the cache cleared first?

**Comment by martindurant** [comment]
As implemented (which is very much POC), the location of temporary files can be specified, or is a temporary directory. If the latter, the same temporary directory path I think gets inherited by all instances in the different workers (this is down to cloudpickle, maybe should be defined), so workers on the same machine would reference the same set of sparse files - but no data would be copied between them. How well it performs to have multiple process doing writing and reading to the same files, I don't know; in theory, the mmap process should be fairly efficient between processes.

On different machines, data on disc might be duplicated if multiple workers access the same parts of the same files, but each worker will start with an empty cache, unless there are shared filesystem resources (which probably perform poorly).

**Comment by martindurant** [comment]
Example:
```
In [1]: import fastparquet as fp

In [2]: from fsspec.implementations.cached import make_caching_class

In [3]: fs = make_caching_class('s3', {})

In [4]: fs.storage
Out[4]: '/var/folders/3d/xxmnsm9j60gfrvffczdt9dsr0000gn/T/tmpavu6td0r'

In [5]: !open /var/folders/3d/xxmnsm9j60gfrvffczdt9dsr0000gn/T/tmpavu6td0r

In [6]: pf = fp.ParquetFile('MDtemp/airline_data_parquet', open_with=fs.open)

In [7]: out = pf.to_pandas(columns=['Year'])

In [8]: !du -sh /var/folders/3d/xxmnsm9j60gfrvffczdt9dsr0000gn/T/tmpavu6td0r
110M	/var/folders/3d/xxmnsm9j60gfrvffczdt9dsr0000gn/T/tmpavu6td0r

In [9]: !ls -l /var/folders/3d/xxmnsm9j60gfrvffczdt9dsr0000gn/T/tmpavu6td0r
total 225584
-rw-r--r--  1 mdurant  staff   81248166 19 Jun 16:26 0898e05c144aa481e959058d0ad44e11455967706a3b18de72f1f79615b2efc3
-rw-r--r--  1 mdurant  staff   76817239 19 Jun 16:26 32c53b3c0a2e017ecc10ac2998eb850eb0c96a37d297283c897c5e57cddba115
-rw-r--r--  1 mdurant  staff  194300512 19 Jun 16:26 39e2a29c33e5f25b6de21230b552e72924ee3be12bfed99f64b568f5c4fd0b0a
-rw-r--r--  1 mdurant  staff   19890837 19 Jun 16:26 411053cd321421210c5f9827fe4b1b9a3c1fad2e20320f257ae413d1b525130c
-rw-r--r--  1 mdurant  staff  131323268 19 Jun 16:26 50af09de5ef77e67f3fa3b640bbca01663ce8a39edfd318f4c38257c097f6dfb
-rw-r--r--  1 mdurant  staff  121112373 19 Jun 16:26 6378703ebd60d593a0af20867a07f99f757cc44f67358f142633d4363ba87ca0
-rw-r--r--  1 mdurant  staff  119922409 19 Jun 16:26 681c037261fede5fd45fdb5f278cd28c2e165ee79b88a1c1080db2b146a84784
-rw-r--r--  1 mdurant  staff  206514609 19 Jun 16:26 7e146ce48fd6180227809b53e98f5a350ddef3528dd58b09e219091055cd8b13
-rw-r--r--  1 mdurant  staff   80621325 19 Jun 16:26 83ba46d31c10e8a6510de9392999aaa6c93820c3aaad6582294f803d3615aff2
-rw-r--r--  1 mdurant  staff   78335293 19 Jun 16:26 88bf6eae581fdc96788505e456910a0d88dce93f831b5fda199bd14b084f7a22
-rw-r--r--  1 mdurant  staff  199757467 19 Jun 16:26 8b2bca6f4bb380e79a178268fdff006722fa4f327d6c73424a06c4698e25cade
-rw-r--r--  1 mdurant  staff      59749 19 Jun 16:25 8dea33194928561a4dce2342caf9e5580c6822b9c86ef22830bcae5afc7198f7
-rw-r--r--  1 mdurant  staff  117313475 19 Jun 16:26 932cce216f0c1748feaabeb21e3b630a33e12e881981f147180dbe8cd5c4bbf6
-rw-r--r--  1 mdurant  staff  191395795 19 Jun 16:26 9818c65b98d4fcdaf9c0ae056c9ec37dce2d39e8d83713f03fd772331405ffe6
-rw-r--r--  1 mdurant  staff  128642249 19 Jun 16:26 b5e7d0a8e232e5991d85656f71bf221c87dbfc81e998735e47d039f44ab9bc03
-rw-r--r--  1 mdurant  staff  118889302 19 Jun 16:26 bd08faa0babae200ee1b72e3397bb9aef508974eda4e0fc454f7b1a6062ecc18
-rw-r--r--  1 mdurant  staff  160240011 19 Jun 16:26 c1a625a16518db8403ed6a72e6ac783a3bbada0f5939b953ca98e48dc2e2ce64
-rw-r--r--  1 mdurant  staff  124296763 19 Jun 16:26 c1bdf395101ee03e076b70f7b1e7e875b1966c4516cb4c8e22763d6e844e49ec
-rw-r--r--  1 mdurant  staff   78414634 19 Jun 16:26 c213dade035841de4b4d4de00d1177116052e8ae4800aec06e23e8dd669073cf
-rw-r--r--  1 mdurant  staff   77998333 19 Jun 16:26 c2de9677af7f01c8944a03a50a75b81c171feda11950e52f0a3d262d93108fa2
-rw-r--r--  1 mdurant  staff       3067 19 Jun 16:26 cache.json
-rw-r--r--  1 mdurant  staff  212673577 19 Jun 16:26 d783cea71f819e6c999904c7ec0c64ff6c5d000877f6c229b89c24049bf723ac
-rw-r--r--  1 mdurant  staff   78498141 19 Jun 16:26 e60dee30b07e8bac41d7cbda6ad17f7011cd5bba68c8f4d11c3497e4288c9143
-rw-r--r--  1 mdurant  staff  120164924 19 Jun 16:26 f292f17deb90c9dc6bba7002d68032989360b44d9a1b46182754ba798e042165
```

i.e., 2.7GB of apparent data only takes up 110MB on disc, because we only accessed little parts of those files.

File: fsspec/implementations/cached.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            else:
+                hash = hashlib.sha256(path.encode()).hexdigest()
+                fn = os.path.join(self.storage, hash)
+                blocks = set()
+                self.cache[path] = {'fn': fn, 'blocks': blocks}
+            f = super()._open(path, mode=mode, cache_type='none', **kwargs)
+            f.cache = MMapCache(f.blocksize, f._fetch_range, f.size,
+                                fn, blocks)
+            close = f.close
+            f.close = lambda: self.close_and_update(f, close)
```
[the remark this answers is not in the document]
Comment (martindurant) [reply fsspec/implementations/cached.py]: Nevermind, must have been something else before, the GC is catching this now.

**Comment by martindurant** [comment]
OK, so problems with this implementation:

- the class is dynamically defined, so you get no instance sharing, and it doesn't pickle (but it does cloudpickle). Also, would be tricky to put in the API docs or to refer to with a "cached:" URL. Could be fixed with clever getattr, or patching the methods of the target filesystem onto the instance, or metaclass hackery
- actually, the cache files will get stomped on if there are multiple instances (in processes) accessing the same cache

**Comment by martindurant** [comment]
@jsignell , @rabernat , I'm tempted to add this to the API and then merge, perhaps put a little blog out about it, so people can try it.

**Comment by jsignell** [comment]
Hmm it'd be nice to have at least a couple tests first.

**Pull request merged**